File: 0155-min-stack/0155-min-stack.java
class MinStack {
    ArrayList<Integer> al;
    int size = 0;

    public MinStack() {
        al = new ArrayList<Integer>();
    }
    
    public void push(int val) {
        al.add(val);
        size++;
    }
    
    public void pop() {
        al.remove(size-1);
        size--;
    }
    
    public int top() {
        return al.get(size-1);
        
    }
    
    public int getMin() {
        // int min = al.get(0);
        // for (int i = 1;i<size;i++)
        // {
        //     if (al.get(i) < min)
        //     {
        //         min = al.get(i);
        //     }
        // }
        // return min;
        return Collections.min(al);
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 0155-min-stack/0155-min-stack.py
class MinStack:

    def __init__(self):
        self.list = []
        self.min = float('inf')

    def push(self, value: int) -> None:
        self.min = min(self.min, value)
        self.list.append([value, self.min])

    def pop(self) -> None:
        self.list.pop()
        if len(self.list) == 0:
            self.min = float('inf')
        else:
            self.min = self.list[-1][1]

    def top(self) -> int:
        return self.list[-1][0]

    def getMin(self) -> int:
        return self.list[-1][1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(value)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()